PathSegments cannot be used with anything else than PathImpl

kkinnunen-apple opened the pull request:
#### d132cc11f0bf78cb0752568e1eb5d707c50ace5a
<pre>
PathSegments cannot be used with anything else than PathImpl
<a href="https://bugs.webkit.org/show_bug.cgi?id=265247">https://bugs.webkit.org/show_bug.cgi?id=265247</a>
<a href="https://rdar.apple.com/118717060">rdar://118717060</a>

Reviewed by Said Abou-Hallawa.

PathImpl and PathSegment were coupled:
PathImpl::appendSegment would call into PathSegment::addToImpl
PathSegment::addToImpl would call into PathImpl

This would be more complex than needed and also limit the PathSegments
to be only used with PathImpl.

Instead:
  - PathSegment is the data that is being held -- no actions toward any
    class that does actual work on the data.
  - PathImpl knows how to use the path segment types: PathImpl::add()
    for each path segment type.
  - PathImpl knows how to use the PathSegment variant:
    PathImpl::addSegment() for the PathSegment, doing generic add() over
    the variants.
  - Make each member function for adding segments PathImpl::add(). This
    way generic algorithms over the variants are consistent to write.
  - The PathImpl::add() parameters are by-value to avoid introducing
    indirection, the arguments are always used.

Construct PathCG, PathCairo from PathStream without applySegments, as
the PathStream segment list can be just iterated.

This works towards being able to play back a PathSegment list to a
CGContext.

* Source/WebCore/platform/graphics/PathImpl.cpp:
(WebCore::PathImpl::appendSegment): Deleted.
* Source/WebCore/platform/graphics/PathImpl.h:
(WebCore::addPathSegment):
* Source/WebCore/platform/graphics/PathSegment.cpp:
(WebCore::PathSegment::addToImpl const): Deleted.
* Source/WebCore/platform/graphics/PathSegment.h:
(WebCore::PathSegment::addTo const):
* Source/WebCore/platform/graphics/PathSegmentData.cpp:
(WebCore::PathMoveTo::addToImpl const): Deleted.
(WebCore::PathLineTo::addToImpl const): Deleted.
(WebCore::PathQuadCurveTo::addToImpl const): Deleted.
(WebCore::PathBezierCurveTo::addToImpl const): Deleted.
(WebCore::PathArcTo::addToImpl const): Deleted.
(WebCore::PathArc::addToImpl const): Deleted.
(WebCore::PathEllipse::addToImpl const): Deleted.
(WebCore::PathEllipseInRect::addToImpl const): Deleted.
(WebCore::PathRect::addToImpl const): Deleted.
(WebCore::PathRoundedRect::addToImpl const): Deleted.
(WebCore::PathDataLine::addToImpl const): Deleted.
(WebCore::PathDataQuadCurve::addToImpl const): Deleted.
(WebCore::PathDataBezierCurve::addToImpl const): Deleted.
(WebCore::PathDataArc::addToImpl const): Deleted.
(WebCore::PathCloseSubpath::addToImpl const): Deleted.
* Source/WebCore/platform/graphics/PathSegmentData.h:
* Source/WebCore/platform/graphics/cairo/PathCairo.cpp:
(WebCore::PathCairo::create):
* Source/WebCore/platform/graphics/cg/PathCG.cpp:
(WebCore::PathCG::create):

Canonical link: <a href="https://commits.webkit.org/271199@main">https://commits.webkit.org/271199@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8b87147434e75a0eebdd6f7012da47bedb4706e6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27331 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5970 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28579 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/29557 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25013 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/27799 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7875 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3369 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24823 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27593 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4767 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23459 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4149 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/4287 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24455 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/30193 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24966 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24874 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/30441 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/4316 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2460 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/28395 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5780 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6641 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4773 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4697 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/platform/graphics/PathImpl.h:
```diff
@@ -50,20 +50,18 @@ class PathImpl : public ThreadSafeRefCounted<PathImpl> {
-    virtual void addRoundedRect(const FloatRoundedRect&, PathRoundedRect::Strategy) = 0;
-
-    virtual void closeSubpath() = 0;
+    void addSegment(PathSegment);
+    virtual void add(PathMoveTo) = 0;
+    virtual void add(PathLineTo) = 0;
+    virtual void add(PathQuadCurveTo) = 0;
+    virtual void add(PathBezierCurveTo) = 0;
+    virtual void add(PathArcTo) = 0;
+    virtual void add(PathArc) = 0;
+    virtual void add(PathEllipse) = 0;
+    virtual void add(PathEllipseInRect) = 0;
+    virtual void add(PathRect) = 0;
+    virtual void add(PathRoundedRect) = 0;
+    virtual void add(PathCloseSubpath) = 0;
```
Is it okay to pass the `PathSegment` and the `PathSegment` data structures by value? `PathRoundedRect` has 12 floats and an enum.

kkinnunen-apple replied:
I would think so -- the PathSegment is only passed when it's a syntactic sugar over the individual calls. So no callframe should be generated, rather the switch should operate on the caller's value.
The individual PathSegment::Data types, especially small ones, should mostly be neutral when passed as a value:
For example, `add(PathMoveTo`) is two floats, where as `moveTo(const FloatPoint&)` passes the two simple values with a pointer indirection. Here the PathMoveTo variant should be a progression.

However, very big ones like `PathRoundedRect`, the ones that might have unread, redundant values, is a bit of a question mark. I would *imagine* that the effect is neutral, but I have not benchmarked it extensively.

Mostly I think the pass-by-value is progression if:
 - There is not many call layers that just forward the parameters (there is not)
 - There is not many variables that are passed redundantly and never read by the payload function (all except rounded rect variant)
 
At least if the problem is the latter, redundant values being passed, this would mean that the command is overloaded: two different methods in one. This is the case with "add rounded rect with platform strategy", "add rounded rect with webkit strategy". So a better, more consistent approach to fix that particular problem would be tahat these two could be split into two separate methods, if they mean different things.

shallawa commented on Source/WebCore/platform/graphics/Path.cpp:
```diff
@@ -152,7 +152,7 @@ void Path::addLineTo(const FloatPoint& point)
         m_data = PathSegment(PathDataLine { moveTo->point, point });
     else
-        ensureImpl().addLineTo(point);
+        ensureImpl().add(PathLineTo { point });
```
If the Impl is `PathCG`, this adds the overhead of copying the point to structure here. Then getting the address of the point in the structure in `PathCG::add(PathMoveTo)`.

So I do not see the benefit of this change so far. Maybe in #[10594](https://github.com/WebKit/WebKit/pull/10594), I will see the benefit of it.

kkinnunen-apple replied:
>If the Impl is PathCG, this adds the overhead of copying the point to structure here. 

No, I don't think there's an additional overhead.
Before, the pointer read to registers happened in PathCG::moveTo.
After,  the pointer read to registers happens in Path::addLineTo.
It's the same amount of reads.

> Then getting the address of the point in the structure in PathCG::add(PathMoveTo).

No, there's no pointers anymore in PathCG::add. It's pass by value.
I think the values are already in the registers.

kkinnunen-apple replied:
> So I do not see the benefit of this change so far. Maybe in #10594, I will see the benefit of it.

The intended benefit of having uniform name and passing the PathSegment::Data values as values:
It is done to be able to write polymorphic functions (==templates) over PathSegment. 

For PathImpl::add over moveTo,bezierCurveTo,arcTo, it's evident: we are able to deconstruct PathSegment and apply the deconstructed variants in uniform fashion, and we don't need to further deconstruct the various PathMoveTo, ..., structs into moveTo(a,b,c,) calls redundantly.

kkinnunen-apple committed:
PathSegments cannot be used with anything else than PathImpl
https://bugs.webkit.org/show_bug.cgi?id=265247
rdar://118717060

Reviewed by Said Abou-Hallawa.

PathImpl and PathSegment were coupled:
PathImpl::appendSegment would call into PathSegment::addToImpl
PathSegment::addToImpl would call into PathImpl

This would be more complex than needed and also limit the PathSegments
to be only used with PathImpl.

Instead:
  - PathSegment is the data that is being held -- no actions toward any
    class that does actual work on the data.
  - PathImpl knows how to use the path segment types: PathImpl::add()
    for each path segment type.
  - PathImpl knows how to use the PathSegment variant:
    PathImpl::addSegment() for the PathSegment, doing generic add() over
    the variants.
  - Make each member function for adding segments PathImpl::add(). This
    way generic algorithms over the variants are consistent to write.
  - The PathImpl::add() parameters are by-value to avoid introducing
    indirection, the arguments are always used.

Construct PathCG, PathCairo from PathStream without applySegments, as
the PathStream segment list can be just iterated.

This works towards being able to play back a PathSegment list to a
CGContext.

* Source/WebCore/platform/graphics/PathImpl.cpp:
(WebCore::PathImpl::appendSegment): Deleted.
* Source/WebCore/platform/graphics/PathImpl.h:
(WebCore::addPathSegment):
* Source/WebCore/platform/graphics/PathSegment.cpp:
(WebCore::PathSegment::addToImpl const): Deleted.
* Source/WebCore/platform/graphics/PathSegment.h:
(WebCore::PathSegment::addTo const):
* Source/WebCore/platform/graphics/PathSegmentData.cpp:
(WebCore::PathMoveTo::addToImpl const): Deleted.
(WebCore::PathLineTo::addToImpl const): Deleted.
(WebCore::PathQuadCurveTo::addToImpl const): Deleted.
(WebCore::PathBezierCurveTo::addToImpl const): Deleted.
(WebCore::PathArcTo::addToImpl const): Deleted.
(WebCore::PathArc::addToImpl const): Deleted.
(WebCore::PathEllipse::addToImpl const): Deleted.
(WebCore::PathEllipseInRect::addToImpl const): Deleted.
(WebCore::PathRect::addToImpl const): Deleted.
(WebCore::PathRoundedRect::addToImpl const): Deleted.
(WebCore::PathDataLine::addToImpl const): Deleted.
(WebCore::PathDataQuadCurve::addToImpl const): Deleted.
(WebCore::PathDataBezierCurve::addToImpl const): Deleted.
(WebCore::PathDataArc::addToImpl const): Deleted.
(WebCore::PathCloseSubpath::addToImpl const): Deleted.
* Source/WebCore/platform/graphics/PathSegmentData.h:
* Source/WebCore/platform/graphics/cairo/PathCairo.cpp:
(WebCore::PathCairo::create):
* Source/WebCore/platform/graphics/cg/PathCG.cpp:
(WebCore::PathCG::create):

Canonical link: https://commits.webkit.org/271199@main
